nixos/modules/services/mail/cyrus-imap.nix

```diff
@@ -0,0 +1,366 @@
+        sasl_pwcheck_method = [ "saslauthd" ];
+        sievedir = "/var/lib/cyrus/sieve";
+        statuscache_db_path = "/run/cyrus/db/statuscache.db";
+        syslog_prefix = "cyrus";
+        tls_client_ca_dir = "/etc/ssl/certs";
+        tls_session_timeout = 1440;
+        tls_sessions_db_path = "/run/cyrus/db/tls_sessions.db";
+        virtdomains = "on";
+      };
+      description = "IMAP configuration settings. [imapd.conf(5)](https://www.cyrusimap.org/imap/reference/manpages/configs/imapd.conf.html)";
+    };
+
+    user = mkOption {
+      type = str;
+      default = "cyrus";
+      description = "Cyrus IMAP user name.";
+    };
+
+    group = mkOption {
+      type = str;
+      default = "cyrus";
+      description = "Cyrus IMAP group name.";
+    };
+
+    imapdConfigFile = mkOption {
+      type = nullOr path;
+      default = null;
+      description = "Config file used for the whole cyrus-imap configuration.";
+      apply = v: if v != null then v else pkgs.writeText "imapd.conf" imapdConfig;
+    };
+
+    cyrusConfigFile = mkOption {
+      type = nullOr path;
+      default = null;
+      description = "Config file used for the whole cyrus-imap configuration.";
+      apply = v: if v != null then v else pkgs.writeText "cyrus.conf" cyrusConfig;
+    };
+
+    sslCACert = mkOption {
+      type = nullOr str;
+      default = null;
+      description = "File containing one or more Certificate Authority (CA) certificates.";
+    };
+
+    sslServerCert = mkOption {
+      type = nullOr str;
+      default = null;
+      description = "File containing the global certificate used for ALL services (imap, pop3, lmtp, sieve)";
+    };
+
+    sslServerKey = mkOption {
+      type = nullOr str;
+      default = null;
+      description = "File containing the private key belonging to the global server certificate.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.users.cyrus = optionalAttrs (cfg.user == "cyrus") {
+      description = "Cyrus IMAP user";
+      isSystemUser = true;
+      group = cfg.group;
+    };
+
+    users.groups.cyrus = optionalAttrs (cfg.group == "cyrus") { };
+
+    environment.etc."imapd.conf".source = cfg.imapdConfigFile;
+    environment.etc."cyrus.conf".source = cfg.cyrusConfigFile;
+
+    systemd.services.cyrus-imap = {
+      description = "Cyrus IMAP server";
+
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      restartTriggers = [
+        cfg.cyrusConfigFile
+        cfg.imapdConfigFile
+      ];
+
+      startLimitIntervalSec = 60; # 1 min
+      environment = {
+        CYRUS_VERBOSE = mkIf cfg.debug "1";
+        LISTENQUEUE = "${builtins.toString cfg.listenQueue}";
+      };
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        Type = "simple";
+        ExecStart = "${cyrus-imapdPkg}/libexec/master -l $LISTENQUEUE -C ${cfg.imapdConfigFile} -M ${cfg.cyrusConfigFile} -p /run/cyrus/master.pid -D";
```

Comment:
If `/etc/imapd.conf` is already unconditionally set to the contents of `cfg.imapdConfigFile`, is this really necessary? Same goes for the cyrus config.